### PR TITLE
Added relationship attribute model with seed data and tests; updated relationship apis to return list of attributes too 

### PR DESCRIPTION
**Pull request by dave-cheong**

### CODE REVIEW

File: backend/typescript/models/base.ts [hunk 1]
```diff
@@ -1,40 +1,62 @@
...
+/* tslint:disable:no-var-requires */
+const mongooseDeepPopulate = require('mongoose-deep-populate')(mongoose);
 
 /**
  * A convenience class to build a query object, only adding criteria when specified.
  */
 export class Query {
 
-    private data:IQueryData = {};
+    private data:IQueryPromise = {};
```
Comment (sundriver) [inline backend/typescript/models/base.ts]: type of data is {[index:string]:Promise<Object>}

File: backend/typescript/models/base.ts [hunk 2]
```diff
@@ -1,40 +1,62 @@
...
 
     /**
      * Adds the given filter (name:value) only if 'condition' is truthy.
      */
-    public add(name:string, value:Object, condition:Object):Query {
+    public when(condition:Object, name:string, callback:() => Promise<Object>) {
```
Comment (sundriver) [inline backend/typescript/models/base.ts]: where do you use condition inside the code ?

Comment (dave-cheong) [inline backend/typescript/models/base.ts]: You don't. Condition if truthy will add the filter clause:

eg

```
            const query = await (new Query()
                .when(subjectIdentityIdValue, 'subject', () => PartyModel.findByIdentityIdValue(subjectIdentityIdValue))
                .when(delegateIdentityIdValue, 'delegate', () => PartyModel.findByIdentityIdValue(delegateIdentityIdValue))
                .build());

```

File: backend/typescript/models/base.ts (hunk 1 above)
Comment (dave-cheong) [inline backend/typescript/models/base.ts]: Fixed. Awesome! Thanks for that. That cleaned up those unnecessary interfaces.

File: backend/typescript/models/relationshipAttribute.model.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+const RelationshipAttributeSchema = new mongoose.Schema({
+    value: {
+      type: String,
+      required: false,
+      trim: true
+    },
+    attributeName: {
+        type: mongoose.Schema.Types.ObjectId,
+        ref: 'RelationshipAttributeName'
```
Comment (sundriver) [inline backend/typescript/models/relationshipAttribute.model.ts]: required?

File: backend/typescript/models/relationshipAttribute.model.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+});
+
+// interfaces .........................................................................................................
+
+export interface IRelationshipAttribute extends mongoose.Document {
+    value?: string;
+    attributeName: IRelationshipAttributeName;
+    toDTO():Promise<DTO>;
```
Comment (sundriver) [inline backend/typescript/models/relationshipAttribute.model.ts]: boolean param

File: backend/typescript/models/relationship.model.ts [hunk 5]
```diff
@@ -95,7 +100,11 @@ const RelationshipSchema = RAMSchema({
         required: [true, 'Status is required'],
         trim: true,
         enum: RelationshipStatus.valueStrings()
-    }
+    },
+    attributes: [{
+        type: mongoose.Schema.Types.ObjectId,
+        ref: 'RelationshipAttribute'
```
Comment (sundriver) [inline backend/typescript/models/relationship.model.ts]: required?

File: backend/typescript/models/relationship.model.ts [hunk 6]
```diff
@@ -110,15 +119,17 @@ export interface IRelationship extends IRAMObject {
...
     statusEnum(): RelationshipStatus;
     toHrefValue(includeValue:boolean):Promise<HrefValue<DTO>>;
     toDTO():Promise<DTO>;
 }
 
 export interface IRelationshipModel extends mongoose.Model<IRelationship> {
-    findByIdentifier:(id:String) => mongoose.Promise<IRelationship>;
-    findPendingByInvitationCodeInDateRange:(invitationCode:String, date:Date) => mongoose.Promise<IRelationship>;
-    search:(subjectIdentityIdValue:string, delegateIdentityIdValue:string, page:number, pageSize:number) => Promise<SearchResult<IRelationship>>;
+    findByIdentifier:(id:String) => Promise<IRelationship>;
```
Comment (sundriver) [inline backend/typescript/models/relationship.model.ts]: mongoose promises are different than es6 promises hence we have mongoose.Promise for all mongoose calls

File: backend/typescript/models/base.ts (hunk 2 above)
Comment (sundriver) [inline backend/typescript/models/base.ts]: ok

File: backend/typescript/models/relationshipAttribute.model.ts (hunk 4 above)
Comment (dave-cheong) [inline backend/typescript/models/relationshipAttribute.model.ts]: Boolean param for? The param include was for the `toHrefValue()` not  the `toDTO()`. There doesn't seem to be hrefs for attributes.

File: backend/typescript/models/relationship.model.ts (hunk 5 above)
Comment (dave-cheong) [inline backend/typescript/models/relationship.model.ts]: Added `required: true`

File: backend/typescript/models/base.ts [hunk 7]
```diff
@@ -1,40 +1,54 @@
...
-}
+    public async build():Promise<Object> {
+        try {
+            const promises = Object.keys(this.data).map(val => this.data[val]);
+            const values:Object[] = await Promise.all<Object>(promises);
+
+            let i = 0;
+            let result:{[name:string]:Object} = {};
+            for (let key of Object.keys(this.data)) {
+                result[key] = values[i];
```
Comment (sundriver) [inline backend/typescript/models/base.ts]: Minor but you can just push into array, do I miss something ?

**Pull request merged**